merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+                            if let Some(other_sequence_number) = value_bucket.sequence_number {
+                                let sequence_diff =
+                                    sequence_number.checked_sub(other_sequence_number);
+                                if let Some(sequence_diff) = sequence_diff {
+                                    if sequence_diff >= self.sequence_threshold {
+                                        *value_bucket = HashSetCell {
+                                            value: value.clone(),
+                                            sequence_number: None,
+                                        };
+                                        self.next_value_index = if self.next_value_index
+                                            < self.capacity_values - 1.into()
+                                        {
+                                            self.next_value_index + I::zero()
+                                        } else {
+                                            0.into()
+                                        };
+                                        return Ok(());
+                                    }
+                                }
+                            }
+                            // Otherwise, we need to prevent having multiple valid
+                            // elements with the same value.
+                            if value_bucket.value == *value {
+                                return Err(HashSetError::ElementAlreadyExists);
+                            }
+                        }
+                        // Panics: If there is a hash set cell pointing to a `None` value,
+                        // it means we really screwed up in the implementation...
+                        // That should never happen.
+                        None => unreachable!(),
+                    }
+                }
+                None => {
+                    let value_bucket = unsafe {
+                        &mut *self.values.as_ptr().add(
+                            usize::try_from(self.next_value_index)
+                                .map_err(|_| HashSetError::UsizeConv)?,
+                        )
+                    };
+                    *index_bucket = Some(self.next_value_index);
+                    *value_bucket = Some(HashSetCell {
+                        value: value.clone(),
+                        sequence_number: None,
+                    });
+                    self.next_value_index =
+                        if self.next_value_index < self.capacity_values - 1.into() {
+                            self.next_value_index + I::one()
+                        } else {
+                            0.into()
+                        };
+                    return Ok(());
+                }
+            }
+        }
+
+        Err(HashSetError::Full)
+    }
+
+    fn find_element(&self, value: &BigUint) -> Result<Option<&mut HashSetCell>, HashSetError> {
```

Comment:
Can we make this public so that we can use it offchain in tests to find an element?

Reply:
Whoops, it's actually public in the code I didn't push yet. :grimacing: Let me push that in 1-2h.